File: web/tests/call-recordings-load.test.ts
// @vitest-environment node
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

/**
 * 🔴 "No calls yet" was the answer to a question the screen never got to ask.
 *
 * GET /api/voice/sessions replies exactly three ways:
 *
 *     200  {ok:true,  sessions:[…]}
 *     401  {ok:false, error:"login required"}
 *     502  {ok:false, error:<the worker's error, or the EDGE's exception text>}
 *
 * Call recordings reached past `Api` to a bare `URLSession`, threw the response
 * away (`let (data, _)`), and decoded into `{ sessions: [CallSession]? }`. An
 * absent key satisfies an optional property, so **both refusal bodies decoded
 * successfully** with `sessions == nil`. The list read `[]` and the screen said,
 * about the user's own recordings, "No calls yet — finished voice calls appear
 * here." Someone whose session had merely expired was told their archive was
 * empty. A worker outage said the same.
 *
 * And on the one path that did report a failure — a true transport error — the
 * caption asserted the connection, in grey body text, with no control to act on.
 *
 * `CallRecordingsLoadTests` (Swift) runs every one of those three answers
 * through the real path with a stubbed transport. This suite owns the wiring:
 * that the status is no longer discarded, that `ok` is not optional, and that
 * the failure state is the house one.
 */

const ROOT = process.cwd()
const VOICE = join(ROOT, 'ios/Tiny/Sources/VoiceCall.swift')
const API = join(ROOT, 'ios/Tiny/Sources/Api.swift')
const ROUTE = join(ROOT, 'app/api/voice/sessions/route.ts')

/** Comments stripped: a rule explained in prose must not satisfy an assertion. */
const code = (src: string) =>
  src.replace(/\/\*[\s\S]*?\*\//g, '').replace(/^\s*(\/\/|\/\/\/).*$/gm, '')

/** The `{ … }` block opening at or after `at`, brace-matched. */
function braced(source: string, at: number): string {
  const open = source.indexOf('{', at)
  let depth = 1
  let i = open + 1
  while (i < source.length && depth > 0) {
    if (source[i] === '{') depth++
    else if (source[i] === '}') depth--
    i++
  }
  return source.slice(open, i)
}

/** Every capture-group-1 match. `[...src.matchAll(re)]` is the obvious spelling
 *  and it typechecks only with `--downlevelIteration`; this tsconfig's target is
 *  below ES2015, so the spread would add two `error TS2802` to a gate that other
 *  suites already carry 30 of. An `exec` loop costs six lines and no noise. */
function caps(source: string, re: RegExp): string[] {
  const g = new RegExp(re.source, re.flags.includes('g') ? re.flags : re.flags + 'g')
  const out: string[] = []
  let m: RegExpExecArray | null
  while ((m = g.exec(source)) !== null) out.push(m[1])
  return out
}

/** A block, with its anchor ASSERTED — an unfound anchor makes `slice` return
 *  one character, on which every `.not.toMatch()` passes forever. */
function body(source: string, signature: string): string {
  const at = source.indexOf(signature)
  expect(at, `${signature} not found — every pin below would be vacuous`).toBeGreaterThan(-1)
  return braced(source, at)
}

describe('call recordings: no answer is not an empty archive', () => {
  it('the load rides the house client, so the status cannot be dropped', () => {
    const load = body(code(readFileSync(VOICE, 'utf8')), 'private func load() async {')
    expect(load, 'the load reaches past Api again — a bare URLSession has no failure contract')
      .not.toMatch(/URLSession\.shared/)
    // `let (data, _)` — the response, discarded. That single underscore is where
    // a 401 and a 502 stopped existing.
    expect(load, 'the response is being discarded again').not.toMatch(/\(\s*data\s*,\s*_\s*\)/)
    expect(load).toMatch(/Api\.getData\("\/api\/voice\/sessions", token: token\)/)
    expect(load, 'the reason is not coming from the shared helper')
      .toMatch(/LoadFailure\.message\(error\)/)
  })

  it('`getData` carries the same failure contract as `get`', () => {
    // It must ride the private `request` core — that is what applies the 30s
    // bound, the Bearer header, and the throw-with-the-server's-words on a
    // non-2xx. A convenience wrapper that called URLSession itself would look
    // identical at the call site and reintroduce the whole defect.
    const fn = body(code(readFileSync(API, 'utf8')), 'static func getData(')
    expect(fn).toMatch(/try await request\(path, token: token\)/)
    expect(fn, 'getData grew its own transport').not.toMatch(/URLSession/)
  })

  it('nothing swallows this decode with `try?` any more', () => {
    // The other half of the defect: even with a status in hand, `try?` on the
    // decode turns a maintenance page into an empty list.
    const src = code(readFileSync(VOICE, 'utf8'))
    const rows = body(src, 'static func rows(from data: Data) throws -> [CallSession] {')
    expect(rows, 'the decode is optional again').not.toMatch(/try\?\s*JSONDecoder/)
    expect(rows).toMatch(/try JSONDecoder\(\)\.decode\(CallSessionsBody\.self/)
    // And the old caption is gone from the file entirely.
    expect(src, 'the screen guesses at the connection again')
      .not.toContain("check your connection.")
  })

  it('`ok` is REQUIRED — an absent key was the whole masked-empty bug', () => {
    // ⚠️ The mechanism, stated once: in Swift an ABSENT key satisfies an
    // OPTIONAL property, so `{ let sessions: [CallSession]? }` decoded the 401
    // body without complaint. A non-optional `ok` cannot be satisfied by a body
    // that isn't the documented success shape.
    const struct = body(code(readFileSync(VOICE, 'utf8')), 'struct CallSessionsBody: Decodable {')
    expect(struct, 'ok went optional — every refusal body decodes again')
      .toMatch(/let ok: Bool\s*$/m)
    expect(struct).not.toMatch(/let ok: Bool\?/)
  })

  it('the rows gate on `ok` and raise the ONE house error', () => {
    const rows = body(code(readFileSync(VOICE, 'utf8')),
                      'static func rows(from data: Data) throws -> [CallSession] {')
    expect(rows).toMatch(/guard body\.ok, let list = body\.sessions else \{ throw ApiError\.badResponse \}/)
    // No sentence of its own: the caption is `LoadFailure`'s job, and a second
    // copy of the status wording is how five sheets drifted apart. Enumerated
    // rather than "no literal over N chars" — that form matched the GAP between
    // `"ended"` and `"error"` (a closing quote can open a match), so it flagged
    // two field values as prose while a real sentence could hide between them.
    const literals = caps(rows, /"([^"]*)"/).sort()
    expect(literals, 'the loader gained a string — a caption belongs in LoadFailure')
      .toEqual(['ended', 'error'])
  })

  it('a live call, a pocket dial and a silent row still stay hidden', () => {
    // Pinned because this increment rewrote the function that holds it, which is
    // the moment an unpinned invariant gets quietly dropped.
    const rows = body(code(readFileSync(VOICE, 'utf8')),
                      'static func rows(from data: Data) throws -> [CallSession] {')
    expect(rows).toMatch(/status == "ended" \|\| \$0\.status == "error"/)
    expect(rows).toMatch(/duration_ms \?\? 0\) > 2000/)
    expect(rows).toMatch(/segment_count \?\? 0\) > 0/)
  })

  it('the failure state offers the remedy it names', () => {
    // It was `Text(errorText).foregroundStyle(.secondary)` — a grey dead end,
    // while the caption it now shows ends in "try again" or "sign out and back
    // in". Every other list sheet in the app uses ContentUnavailableView with a
    // Retry; this one is the house shape now too.
    const src = code(readFileSync(VOICE, 'utf8'))
    const at = src.indexOf('} else if let errorText {')
    expect(at, 'the failure branch moved — re-anchor').toBeGreaterThan(-1)
    const branch = src.slice(at, src.indexOf('} else if sessions.isEmpty {', at))
    expect(branch.length, 'read the wrong region').toBeGreaterThan(80)
    expect(branch).toMatch(/ContentUnavailableView/)
    expect(branch).toMatch(/Text\(errorText\)/)
    expect(branch).toMatch(/Button\("Retry"\)/)
    expect(branch, 'a grey sentence with nothing to do about it')
      .not.toMatch(/Text\(errorText\)\.foregroundStyle\(\.secondary\)/)
  })

  it('the failure glyph does not name a cause the app never checked', () => {
    // ⚠️ A crossed-out wifi symbol over "Session expired — sign out and back in"
    // blames the wrong thing, and the picture is what gets read first. The house
    // pattern that IS honest is the screen's own subject, crossed out
    // (bolt.slash on Activity, iphone.slash on My Devices) — it says "this
    // content isn't here", which is all the app knows.
    const src = code(readFileSync(VOICE, 'utf8'))
    const at = src.indexOf('} else if let errorText {')
    const branch = src.slice(at, src.indexOf('} else if sessions.isEmpty {', at))
    expect(branch).toMatch(/Label\("Couldn't load calls", systemImage: "waveform\.slash"\)/)
    for (const cause of ['wifi.slash', 'antenna.radiowaves.left.and.right.slash', 'network.slash']) {
      expect(branch, `the glyph asserts ${cause} for a failure that is usually a 401`)
        .not.toContain(cause)
    }
  })

  it('the empty state is still allowed to say empty', () => {
    // The fix is not "never say empty" — a real 200 with no rows must still
    // reach the empty state, or the screen just moves the lie.
    const src = code(readFileSync(VOICE, 'utf8'))
    expect(src).toMatch(/ContentUnavailableView\("No calls yet"/)
    // And it sits AFTER the error branch, so a failure can never fall into it.
    expect(src.indexOf('"No calls yet"')).toBeGreaterThan(src.indexOf('} else if let errorText {'))
  })

  // ── the third surface ─────────────────────────────────────────────────────

  /**
   * Android told the same lie, and got there by a worse road.
   *
   * Its sheet also reached past the house client — to a bare `HttpURLConnection` —
   * but `conn.inputStream` THROWS on a 401 or a 502 (that is `getErrorStream`'s
   * job). So `runCatching { … }.getOrNull()` flattened BOTH refusals into one
   * sentence, "Couldn't load calls — check your connection", which on an expired
   * session blames the network for the app's own state and points at a remedy that
   * cannot work. Worse than iOS's: iOS at least distinguished nothing, quietly;
   * Android asserted a wrong cause.
   *
   * It also set `calls = emptyList()` on failure while ordering the spinner arm
   * before the error arm — so the empty state was reachable on a refusal, which is
   * the "No calls yet" claim itself.
   *
   * Kotlin `CallRecordingsLoadTest` owns the rule (all three answers, no network).
   * These pins own the wiring.
   */
  const KT = join(ROOT, 'android/app/src/main/java/technology/tiny/app/ui/CallRecordingsSheet.kt')
  /** Kotlin comments stripped — a rule explained in prose must not satisfy a pin. */
  const kt = () =>
    readFileSync(KT, 'utf8')
      .replace(/\/\*[\s\S]*?\*\//g, '')
      .replace(/^\s*\/\/.*$/gm, '')

  it('the Android load rides the house client too', () => {
    const src = kt()
    // ⚠️ THE DEFECT, in one regex. The bypass is what threw the status away, and a
    // screen with no status can only guess at a cause.
    expect(src, 'the sheet reaches past app.api again — a raw connection has no failure contract')
      .not.toMatch(/HttpURLConnection/)
    expect(src, 'a raw URL fetch is back').not.toMatch(/URL\(\s*"\$\{app\.config\.serverBase\}/)
    expect(src).toMatch(/app\.api\.getJson\("\/api\/voice\/sessions"\)/)
    // And the old single caption, which asserted a cause for two answers that
    // weren't it.
    expect(src, 'the connection is being blamed for every refusal again')
      .not.toMatch(/Couldn't load calls — check your connection/)
  })

  it('the Android sheet asks the shared rule for both halves', () => {
    const src = kt()
    expect(src, 'the rows are no longer coming from the checked rule')
      .toMatch(/CallRecordingsLoad\.rows\(res\)/)
    expect(src, 'the caption is no longer the shared rule\'s')
      .toMatch(/error = CallRecordingsLoad\.message\(res\)/)
    // The rule delegates to the same LoadFailure the other six sheets use, rather
    // than growing a seventh copy of the status wording.
    expect(src).toMatch(/LoadFailure\.loaded\(res, "sessions"\)/)
    expect(src, 'the rule writes its own status sentence')
      .not.toMatch(/friendlyHttpError/)
  })

  it('a failed Android load cannot reach the empty state', () => {
    // ⚠️ The ordering IS the bug. `calls` stays null on failure, so if the spinner
    // arm came first a refusal would spin forever — and if `calls` were set to an
    // empty list, the empty state would say "No calls yet" about an answer that
    // never came. Both were true before.
    const src = kt()
    const errAt = src.indexOf('error != null ->')
    const loadingAt = src.indexOf('calls == null ->')
    const emptyAt = src.indexOf('calls!!.isEmpty() ->')
    expect(errAt, 'the failure arm went away — re-anchor').toBeGreaterThan(-1)
    expect(loadingAt, 'the loading arm went away — re-anchor').toBeGreaterThan(-1)
    expect(emptyAt, 'the empty arm went away — re-anchor').toBeGreaterThan(-1)
    expect(errAt, 'the failure arm no longer precedes the spinner — a refusal spins forever')
      .toBeLessThan(loadingAt)
    expect(loadingAt, 'the spinner no longer precedes the empty state').toBeLessThan(emptyAt)
    // And no failure path hands the empty state a list to be confident about.
    const eff = body(src, 'LaunchedEffect(reloadKey) {')
    expect(eff, 'a failed load sets an empty list again — that IS the "No calls yet" lie')
      .not.toMatch(/calls = emptyList\(\)/)
  })

  it('the Android failure state offers the remedy it names', () => {
    // Grey body text with no control was a dead end, and the caption now often ends
    // in "try again". The house shape (Jobs, My Devices): reason + retry.
    const src = kt()
    const at = src.indexOf('error != null ->')
    expect(at).toBeGreaterThan(-1)
    const arm = src.slice(at, loadingArmStart(src, at))
    expect(arm, 'the failure arm stopped showing the reason').toMatch(/Text\(error!!/)
    expect(arm, 'the failure arm offers no way out').toMatch(/TextButton\(/)
    expect(arm, 'the retry does not re-run the load').toMatch(/reloadKey\+\+/)
    // The load must actually be keyed on it, or the button is decoration.
    expect(src, 'the retry key does not drive the load').toMatch(/LaunchedEffect\(reloadKey\)/)
  })

  /** The failure arm ends where the loading arm begins — scoped, so a pin on the
   *  failure arm cannot be satisfied by a control in a later branch. */
  function loadingArmStart(src: string, from: number): number {
    const next = src.indexOf('calls == null ->', from)
    expect(next, 'the loading arm no longer follows the failure arm').toBeGreaterThan(from)
    return next
  }

  it('the Android empty state is still allowed to say empty', () => {
    // The fix is not "never say empty" — a real 200 with no rows must still reach
    // it, or the screen just moves the lie.
    expect(kt()).toMatch(/No calls yet/)
  })

  it('⚠️ FAILS WHEN FIXED: the route answers only 401 and 502', () => {
    // The caption is only as good as the table's coverage. 401 and 5xx are both
    // statuses the table OWNS (`statusOwnsTheMessage`), which is what keeps the
    // edge's raw "The operation was aborted due to timeout" off the screen. A
    // 402 or 404 here would be worse than a bare code: the table words those for
    // CHAT ("That tiny doesn't exist"), which on a recordings list is a lie.
    const src = readFileSync(ROUTE, 'utf8')
    const statuses = caps(src, /,\s*(\d{3})\s*\)/).map(Number)
    expect(statuses.sort(), '🎉 the route gained a status — check the table has words for it')
      .toEqual([401, 502])
    // And the premise the optional `sessions` rests on: a 200 ALWAYS carries the
    // key, so an absent one really is a broken body rather than "none".
    expect(src).toMatch(/json\(\{\s*ok:\s*true,\s*sessions:\s*res\?\.sessions \|\| \[\]\s*\}\)/)
  })
})

File: android/app/src/test/java/technology/tiny/app/ui/CallRecordingsLoadTest.kt
package technology.tiny.app.ui

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.json.JSONArray
import org.json.JSONObject
import technology.tiny.app.net.friendlyHttpError

/**
 * 🔴 "No calls yet" was said to people whose session had merely expired, and again
 * during a worker outage. iOS `3eca0cfe` ported — Android reached the same lie by a
 * worse road.
 *
 * GET /api/voice/sessions answers exactly three ways:
 *   200 `{ok:true, sessions:[…]}`
 *   401 `{ok:false, error:"login required"}`
 *   502 `{ok:false, error:…}`   (worker unreachable, or it answered an error)
 *
 * iOS decoded all three into a struct whose `sessions` was OPTIONAL, so both
 * refusals decoded cleanly as nil → `[]` → the empty state. Android's sheet instead
 * reached past `app.api` to a bare `HttpURLConnection`, and `conn.inputStream`
 * THROWS on a 401 or a 502 — so `runCatching { … }.getOrNull()` flattened every
 * refusal into ONE sentence, "Couldn't load calls — check your connection". Two
 * distinct server answers, one caption, and it blamed the network for an expired
 * session — sending the reader at a remedy that cannot work.
 *
 * The load now goes through `app.api.getJson`, which keeps the code as `_status`,
 * and the decode is split out here so all three answers are checkable without a
 * network. That split is the point: the old shape made the defect untestable.
 */
class CallRecordingsLoadTest {

    /** One row as the worker's VOICE_LIST_SQL returns it. */
    private fun row(
        id: String = "s1",
        status: String = "ended",
        durationMs: Long = 30_000,
        segments: Long = 4,
        tiny: String? = "ada",
    ) = JSONObject()
        .put("id", id)
        .put("status", status)
        .put("duration_ms", durationMs)
        .put("segment_count", segments)
        .apply { if (tiny != null) put("tiny_name", tiny) }

    /** A real 200 from the route. */
    private fun ok(vararg rows: JSONObject) = JSONObject()
        .put("ok", true)
        .put("sessions", JSONArray().apply { rows.forEach { put(it) } })

    /** What `executeJson` hands back for the route's 401 / 502. */
    private fun refusal(status: Int, error: String) = JSONObject()
        .put("ok", false)
        .put("error", error)
        .put("_status", status)

    // ── the defect: a refusal is not an empty archive ────────────────────────

    @Test
    fun `an expired session is not an empty archive`() {
        // The exact 401 body, and the whole point: NO rows come back, so the sheet
        // cannot reach its empty state and say "No calls yet" about someone's own
        // recordings.
        val res = refusal(401, "login required")
        assertNull("a 401 yielded rows — the empty state would claim the archive is empty",
                   CallRecordingsLoad.rows(res))
        assertNotNull(CallRecordingsLoad.message(res))
    }

    @Test
    fun `a worker outage is not an empty archive either`() {
        val res = refusal(502, "voice worker unreachable")
        assertNull(CallRecordingsLoad.rows(res))
        assertNotNull(CallRecordingsLoad.message(res))
    }

    @Test
    fun `the two refusals no longer share one caption`() {
        // ⚠️ THE ANDROID DEFECT. The bypass collapsed 401 and 502 into the same
        // sentence, so the screen could not tell an expired session from an outage —
        // and the remedies are opposite ones.
        val expired = CallRecordingsLoad.message(refusal(401, "login required"))!!
        val outage = CallRecordingsLoad.message(refusal(502, "voice worker unreachable"))!!
        assertFalse("401 and 502 still read the same: $expired", expired == outage)
        // And neither blames the connection, which is what the old single line did.
        assertFalse("still blames the connection: $expired", expired.contains("connection"))
        // The words come from the shared table, not a seventh copy.
        assertEquals(friendlyHttpError(401), expired)
        assertEquals(friendlyHttpError(502), outage)
    }

    @Test
    fun `nothing answered still says so, and says it once`() {
        // A thrown request (airplane mode, DNS) really is the connection — and that
        // is the ONE case the old sentence was right about. It keeps the house line,
        // byte-shared with the other six sheets.
        assertNull(CallRecordingsLoad.rows(null))
        assertEquals(LoadFailure.noResponse, CallRecordingsLoad.message(null))
        // Not two causes in one breath — the defect iOS's own commit named.
        assertFalse(CallRecordingsLoad.message(null)!!.contains(" or network"))
    }

    @Test
    fun `a 200 that was not JSON is not an empty archive`() {
        // The other half of the collapse (e24f07bf): a 200 whose body isn't JSON
        // becomes an empty JSONObject with no `_status` at all. Shape check, not
        // status check, is what catches it.
        val nonJson200 = JSONObject()
        assertNull(CallRecordingsLoad.rows(nonJson200))
        val m = CallRecordingsLoad.message(nonJson200)!!
        assertTrue("does not name what failed: $m", m.contains("your call recordings"))
        assertFalse("blames the connection for a body that arrived: $m", m.contains("connection"))
    }

    @Test
    fun `a 200 whose body says otherwise is a failure`() {
        // `ok` is checked, not assumed. An intermediary between the app and the
        // worker is exactly what pairs a 200 with a body that refuses — iOS made
        // `ok` non-optional for this; Android has to ask.
        val res = JSONObject().put("ok", false).put("sessions", JSONArray())
        assertNull("a 2xx saying ok:false was read as a good, empty load",
                   CallRecordingsLoad.rows(res))
        // And the wording, not merely that one exists: this is the ONE path that
        // reaches the `?:` fallback, so nothing else pins its sentence. A mutant that
        // made it `unusableBody("it")` survived the first run — "couldn't read it" on
        // a screen full of recordings names nothing at all.
        val m = CallRecordingsLoad.message(res)!!
        assertEquals(LoadFailure.unusableBody("your call recordings"), m)
        assertTrue("does not name what failed: $m", m.contains("your call recordings"))
    }

    // ── and the fix must not just move the lie ───────────────────────────────

    @Test
    fun `a real empty archive is still allowed to say empty`() {
        // The fix is NOT "never say empty". A genuine 200 with no rows must still
        // reach the empty state, or the screen just relocates the lie.
        val rows = CallRecordingsLoad.rows(ok())
        assertNotNull(rows)
        assertTrue(rows!!.isEmpty())
        assertNull("a good empty load produced a caption", CallRecordingsLoad.message(ok()))
    }

    @Test
    fun `a real archive comes back mapped`() {
        val rows = CallRecordingsLoad.rows(ok(row(id = "a"), row(id = "b", tiny = null)))!!
        assertEquals(listOf("a", "b"), rows.map { it.id })
        assertEquals("ada", rows[0].tiny)
        // A row with no name is still playable — "tiny" rather than a blank button.
        assertEquals("tiny", rows[1].tiny)
        assertEquals(30_000L, rows[0].durationMs)
    }

    // ── the filter, which the port had to carry across unchanged ─────────────

    @Test
    fun `only finished calls with real audio are listed`() {
        // Live calls 409 on stitch; sub-2s pocket dials and zero-segment rows have no
        // audio journaled at all (outage casualties — their stitch 404s). Each
        // exclusion is a row that would render a play button that cannot play.
        val res = ok(
            row(id = "live", status = "live"),
            row(id = "short", durationMs = 1_500),
            row(id = "silent", segments = 0),
            row(id = "good"),
            row(id = "errored", status = "error"),
        )
        val ids = CallRecordingsLoad.rows(res)!!.map { it.id }
        assertEquals(listOf("good", "errored"), ids)
    }

    @Test
    fun `a boundary duration is excluded, not rounded in`() {
        // `> 2_000`, not `>=`: exactly 2s is still a pocket dial. Pinned because a
        // comparison is the easiest thing to loosen by one character.
        assertTrue(CallRecordingsLoad.rows(ok(row(durationMs = 2_000)))!!.isEmpty())
        assertEquals(1, CallRecordingsLoad.rows(ok(row(durationMs = 2_001)))!!.size)
    }

    @Test
    fun `a malformed row is skipped, not fatal`() {
        // The array is the server's; one bad element must not lose the calls around
        // it. `getJSONObject` used to throw here, inside the runCatching — so a
        // single junk row emptied the whole archive AND blamed the connection.
        val arr = JSONArray().apply {
            put(row(id = "first"))
            put("not an object")
            put(row(id = "last"))
        }
        val res = JSONObject().put("ok", true).put("sessions", arr)
        assertEquals(listOf("first", "last"), CallRecordingsLoad.rows(res)!!.map { it.id })
    }

    @Test
    fun `a sessions key that is not an array is a failure, not an empty archive`() {
        // ⚠️ `has("sessions")` and "sessions is a list" are DIFFERENT questions, and
        // both survivors of the first mutation run lived in the gap. The route does
        // `sessions: res?.sessions || []` — the worker's value passes straight
        // through — so an object or a string arrives as a 200 with the key present.
        // The shape check says yes; `optJSONArray` still returns null.
        for (bad in listOf(JSONObject().put("count", 3), "none", 7)) {
            val res = JSONObject().put("ok", true).put("sessions", bad)
            assertNull("a non-array `sessions` ($bad) read as an empty archive",
                       CallRecordingsLoad.rows(res))
            val m = CallRecordingsLoad.message(res)
            assertNotNull("a non-array `sessions` ($bad) produced no caption", m)
            // The body arrived and parsed, so the caption must name the subject
            // rather than blame the connection or show a bare 200.
            assertTrue("does not name what failed: $m", m!!.contains("your call recordings"))
            assertFalse("blames the connection: $m", m.contains("connection"))
        }
    }

    @Test
    fun `an absent ok is not taken for consent`() {
        // `optBoolean("ok")` defaults to FALSE, and that is deliberate: a 200 whose
        // body never claimed success isn't the documented shape, and defaulting the
        // other way makes every unrecognised body a good load. Survived until pinned.
        val res = JSONObject().put("sessions", JSONArray())
        assertNull(CallRecordingsLoad.rows(res))
        assertNotNull(CallRecordingsLoad.message(res))
    }

    @Test
    fun `a body with no sessions key at all is a failure, not an empty list`() {
        // An absent key is not an empty array. This is the shape that made two of the
        // route's three answers read as "your recordings are gone".
        val res = JSONObject().put("ok", true)
        assertNull(CallRecordingsLoad.rows(res))
        assertNotNull(CallRecordingsLoad.message(res))
    }

    // ── Why the call ended ───────────────────────────────────────────────────
    // 🔴 `voice_sessions.error` carries a reason for every abnormal end, and the
    // worker's own docstring says what for: "the row is what the person still has
    // tomorrow when they ask why." [CallRecording] had no such field, so the reason
    // arrived and was dropped one line before the render — and because [rows]
    // deliberately ADMITS `status == "error"` rows, a call the voice service killed
    // 20 seconds in drew `📞 ada · 0:20`, identical to a 20-second call the person
    // ended themselves. The duration is what makes them indistinguishable.

    @Test
    fun `a dropped call no longer reads like a short one`() {
        val res = ok(
            row(id = "dropped", durationMs = 20_000).put("error", "upstream closed: 1011 going away"),
            row(id = "clean", durationMs = 20_000),
        )
        val rows = CallRecordingsLoad.rows(res)!!
        // The badge must not become a filter — both calls are the person's.
        assertEquals(listOf("dropped", "clean"), rows.map { it.id })
        assertEquals("the voice service closed the connection", rows[0].outcome)
        // An ordinary hangup gets nothing: a badge on every row says nothing.
        assertNull(rows[1].outcome)
    }

    @Test
    fun `the worker-tail diagnostic never reaches the person`() {
        // ⚠️ The reason is written for the worker tail — a close code, or an
        // arbitrary upstream exception message. Keeping the column diagnostic is
        // right; painting it onto someone's call list would be the same
        // wrong-surface mistake pointing the other way.
        val closed = CallOutcome.text("ended", "upstream closed: 1011 going away")!!
        assertFalse(closed.contains("1011"))
        assertFalse(closed.contains("going away"))
        val errored = CallOutcome.text("error", "upstream error: TypeError: x is not a function")
        assertEquals("the voice service dropped", errored)
        assertFalse(errored!!.contains("TypeError"))
    }

    @Test
    fun `the arms that measured their own cause say what they measured`() {
        assertEquals("we stopped hearing this device", CallOutcome.text("ended", "the client went silent"))
        assertEquals("the call hit the maximum length", CallOutcome.text("ended", "the call hit the maximum length"))
        assertEquals("this device's connection dropped", CallOutcome.text("error", "the client socket errored"))
    }

    @Test
    fun `an unrecognised reason still says the call broke, and names no cause`() {
        // A teardown arm added upstream, or a row from a build this map predates.
        // ⚠️ Falling back to SILENCE here is the original defect returning: the
        // reason would once again have no reader. UNKNOWN is the honest middle.
        val out = CallOutcome.text("ended", "the flux capacitor desynced")
        assertEquals(CallOutcome.UNKNOWN, out)
        assertFalse("it echoed the diagnostic it did not understand", out!!.contains("flux"))
    }

    @Test
    fun `an error row from before the reason was wired is not treated as clean`() {
        // Every error row written before the wiring has status='error', error=NULL.
        // ⚠️ And the JSON path matters as much as the function: `optString` returns
        // "" for a JSON null AND for an absent key, so both legacy shapes must land
        // on the status check rather than reading as a clean hangup.
        assertEquals(CallOutcome.UNKNOWN, CallOutcome.text("error", null))
        val rows = CallRecordingsLoad.rows(ok(row(id = "legacy", status = "error")))!!
        assertEquals(CallOutcome.UNKNOWN, rows.single().outcome)
    }

    @Test
    fun `an ordinary call is not badged`() {
        assertNull(CallOutcome.text("ended", null))
        assertNull(CallOutcome.text("ended", ""))
        // ⚠️ Whitespace is not a reason: `"upstream closed: "` with an empty code
        // and reason trims to a bare prefix, and a row of blanks must not raise a
        // warning on a call that ended fine.
        assertNull(CallOutcome.text("ended", "   "))
        assertNull(CallRecordingsLoad.rows(ok(row()))!!.single().outcome)
    }

    // ── 🔇 Why a recording won't play (lib/voice/playback.ts's Kotlin twin) ──
    // `/voice/recording/:id` STITCHES a call's PCM on first listen and can decline
    // six ways, each a JSON body with a stated reason. This sheet registered
    // `setOnPreparedListener` and `setOnCompletionListener` and NO
    // `setOnErrorListener` — and because `prepareAsync` reports asynchronously,
    // the enclosing `runCatching` could not see it: nothing throws. The row was
    // left mid-play forever, its reason unread.

    @Test
    fun `a refusal always says something`() {
        // The difference from CallOutcome.text, which returns null for a clean call:
        // this is called ONLY when a play failed, so silence is the frozen row again.
        for (input in listOf(null, "", "   ", "who knows")) {
            assertTrue("a failed play said nothing for ${input}",
                       CallRecordingRefusal.text(input).isNotEmpty())
        }
    }

    @Test
    fun `each refusal the route gives has its own sentence`() {
        assertTrue(CallRecordingRefusal.text("call still in progress").contains("still going"))
        assertTrue(CallRecordingRefusal.text("call too long to stitch").contains("too long"))
        assertTrue(CallRecordingRefusal.text("no replay journaled for this session")
                       .contains("wasn't recorded"))
        assertTrue(CallRecordingRefusal.text("no audio journaled").contains("audio wasn't saved"))
        assertTrue(CallRecordingRefusal.text("media store not provisioned")
                       .contains("unavailable right now"))
        // None of the five falls back to the generic line — that would be a covered
        // refusal rendered as an unknown one.
        for (r in listOf("call still in progress", "call too long to stitch",
                         "no replay journaled for this session", "no audio journaled",
                         "media store not provisioned")) {
            assertFalse("$r rendered as the generic sentence",
                        CallRecordingRefusal.text(r) == CallRecordingRefusal.UNKNOWN)
        }
    }

    @Test
    fun `a wrapped reason is still recognised`() {
        // ⚠️ MediaPlayer never hands over a body at all — `onError` carries two ints.
        // `text` matches by CONTAINS so that the one channel which can carry a reason
        // (a status probe, or a platform that embeds the origin's body) still lands on
        // the right sentence. An equality match would pass every test written with
        // bare strings and drop every real refusal.
        val wrapped = "server said: call too long to stitch"
        assertTrue(CallRecordingRefusal.text(wrapped).contains("too long"))
    }

    @Test
    fun `an unreadable refusal names no cause`() {
        val out = CallRecordingRefusal.text("the flux capacitor desynced")
        assertEquals(CallRecordingRefusal.UNKNOWN, out)
        assertFalse("the raw diagnostic reached the person", out.contains("flux"))
    }

    @Test
    fun `the size refusal is knowable before the tap`() {
        // (n - 2) * 1_440_000 > 40_000_000 — two segments exempt because
        // segment_count sums both directions and only the final segment per
        // direction may be short.
        assertTrue("30 segments cannot stitch", CallRecordingRefusal.tooLong(30))
        assertFalse("29 may still stitch", CallRecordingRefusal.tooLong(29))
        // One-sided: a small count is never a promise the call WILL play.
        for (n in listOf(0L, 1L, 2L, 4L)) {
            assertFalse("count $n read as too long", CallRecordingRefusal.tooLong(n))
        }
    }

    @Test
    fun `the segment count survives the decode as a number`() {
        // ⚠️ All three clients read `segment_count` and used it ONLY as `> 0`, so the
        // number that predicts the 413 was on the row and thrown away at the decode.
        // A `Boolean hasAudio` field here would pass every other test in this file.
        val rows = CallRecordingsLoad.rows(ok(row(id = "long", segments = 30)))!!
        assertEquals(30L, rows.single().segmentCount)
        assertTrue(CallRecordingRefusal.tooLong(rows.single().segmentCount))
    }

    @Test
    fun `rows and caption never both exist, and never both miss`() {
        // The two halves cannot disagree: exactly one of "here are the rows" and
        // "here is why not" is true for every answer the route can give.
        val answers = listOf(
            null,
            JSONObject(),
            refusal(401, "login required"),
            refusal(502, "worker down"),
            JSONObject().put("ok", false).put("sessions", JSONArray()),
            JSONObject().put("ok", true),
            ok(),
            ok(row()),
        )
        for (res in answers) {
            val rows = CallRecordingsLoad.rows(res)
            val msg = CallRecordingsLoad.message(res)
            assertTrue("rows and caption disagree for $res", (rows == null) == (msg != null))
        }
    }
}
